File: go/rc4.go
package main

import (
	"fmt"
	"time"
)

type RC4 struct {
	s [256]byte
	i int
	j int
}

func (rc4 *RC4) getByte() byte {
	rc4.i = (rc4.i + 1) % 256
	rc4.j = (rc4.j + int(rc4.s[rc4.i])) % 256

	a := rc4.s[rc4.i]
	b := rc4.s[rc4.j]
	rc4.s[rc4.i] = b
	rc4.s[rc4.j] = a

	idx := (int(rc4.s[rc4.i]) + int(rc4.s[rc4.j])) % 256
	return rc4.s[idx]
}

func newRc4(key []byte) RC4 {
	data := new([256]byte)

	for i := 0; i < 256; i++ {
		data[i] = byte(i)
	}

	j := 0
	for i := 0; i < 256; i++ {
		j = (j + int(data[i]) + int(key[i%len(key)])) % 256
		a := data[i]
		b := data[j]
		data[i] = b
		data[j] = a
	}

	rc4 := RC4{
		*data,
		0, 0,
	}
	return rc4
}

func bench() {
	start := time.Now()

	key := []byte("Keyfobsrulethebestofall")
	rc4 := newRc4(key)

	sum := uint64(0)
	for i := 0; i < 250000000; i++ {
		sum += uint64(rc4.getByte())
	}

	// gccgo does not suport Milliseconds()
	delta := int(time.Since(start).Nanoseconds() / 1000000)
	fmt.Printf("|go|rc4|%d|%+v|\n", sum, delta)
}

func main() {
	bench()
}

File: zig/rc4.zig
const std = @import("std");
const print = std.debug.print;

pub const RC4 = struct {
    s: [256]u8,
    i: u64,
    j: u64,

    pub fn new(key: []const u8) RC4 {
        const state = RC4.key_schedule(key);
        return RC4{ .s = state, .i = 0, .j = 0 };
    }

    fn key_schedule(key: []const u8) [256]u8 {
        var data: [256]u8 = undefined;
        for (data) |*ref, i| {
            ref.* = @intCast(u8, i);
        }

        var j: u32 = 0;
        for (data) |_, i| {
            j = @intCast(u32, j + data[i] + key[i % key.len]) % 256;
            const a = data[i];
            const b = data[j];
            data[j] = a;
            data[i] = b;
        }
        return data;
    }

    fn get_byte(self: *RC4) u8 {
        self.i = (self.i + 1) % 256;
        self.j = (self.j + self.s[self.i]) % 256;

        const a = self.s[self.i];
        const b = self.s[self.j];
        self.s[self.j] = a;
        self.s[self.i] = b;

        const idx = (@intCast(u32, self.s[self.i]) + self.s[self.j]) % 256;
        return self.s[idx];
    }
};

export fn bench() void {
    var start = std.time.milliTimestamp();

    const key = "Keyfobsrulethebestofall"[0..];
    var rc4 = RC4.new(key);
    var sum: u64 = 0;
    var i: u32 = 0;
    while (i < 250000000) : (i += 1) {
        sum += rc4.get_byte();
    }

    const end = std.time.milliTimestamp();
    const delta = end - start;
    print("|zig|rc4|{}|{}|\n", .{ sum, delta });
}

export fn prod(a: i64, b: i64) i64 {
    return a * b;
}

pub fn main() void {
    bench();
}
